File: apps/fsm-core-db-py/src/fsm_core_db/fsm_helper.py
import json
from typing import Any, Optional
import asyncpg

from .constants import FSM_SCHEMA, FSM_SCHEMA_FN_VERSION
from .pg_utils import to_jsonb_param

_V = FSM_SCHEMA_FN_VERSION

MICROSTEP_FN = f"{FSM_SCHEMA}.microstep_{_V}"
SELECT_TRANSITIONS_FN = f"{FSM_SCHEMA}.select_all_transitions_{_V}"
LOAD_FSM_STATE_FN = f"{FSM_SCHEMA}.load_fsm_state_from_json_{_V}"
LOAD_FSM_TRANSITION_FN = f"{FSM_SCHEMA}.load_fsm_transition_from_json_{_V}"
LOAD_FSM_FROM_JSON_FN = f"{FSM_SCHEMA}.load_fsm_from_json_{_V}"
RESOLVE_STATE_VALUE_FN = f"{FSM_SCHEMA}.resolve_state_value_{_V}"


async def select_transitions(
    pool: asyncpg.Pool,
    event_name: str,
    source_state_value_set: Any,
    fsm_name: str,
    fsm_version: str,
) -> list[dict]:
    """
    Select all valid transitions for the given event and current state nodes.
    Mirrors selectTransitions() in fsm-helper.ts.
    """
    sql = f"""
        SELECT * FROM {SELECT_TRANSITIONS_FN}(
            $1::text, $2::jsonb, $3::text, $4::text
        );
    """
    async with pool.acquire() as conn:
        rows = await conn.fetch(
            sql, event_name, to_jsonb_param(source_state_value_set), fsm_name, fsm_version
        )
    return [dict(r) for r in rows]


async def perform_microstep(
    pool: asyncpg.Pool,
    transition_record: Optional[dict],
    event_name: str,
    state_value_node_set: Any,
    fsm_name: str,
    fsm_version: str,
) -> dict:
    """
    Execute a single microstep for the FSM, returning updated state and actions.
    Mirrors performMicrostep() in fsm-helper.ts.
    """
    transition_json = json.dumps(transition_record) if transition_record is not None else None
    node_array = list(state_value_node_set) if isinstance(state_value_node_set, (list, set)) else []
    sql = f"""
        SELECT {MICROSTEP_FN}(
            CASE
                WHEN $1::jsonb IS NULL THEN NULL::{FSM_SCHEMA}.fsm_transitions
                ELSE jsonb_populate_record(NULL::{FSM_SCHEMA}.fsm_transitions, $1::jsonb)
            END,
            $2::text,
            $3::text[],
            $4::text,
            $5::text
        ) AS result;
    """
    async with pool.acquire() as conn:
        row = await conn.fetchrow(sql, transition_json, event_name, node_array, fsm_name, fsm_version)
    if row and row["result"] is not None:
        result = row["result"]
        return dict(result) if hasattr(result, "keys") else result
    return {
        "updated_state_value_node_set": [],
        "updated_state_value": None,
        "exit_actions": None,
        "entry_actions": None,
        "initial_actions": None,
        "transition_actions": None,
    }


async def load_fsm_state_from_json(
    pool: asyncpg.Pool, fsm_name: str, fsm_version: str, state_json: Any
) -> Optional[dict]:
    """Mirrors loadFsmStateFromJsonV2() in fsm-helper.ts."""
    sql = f"SELECT * FROM {LOAD_FSM_STATE_FN}($1::text, $2::text, $3::jsonb);"
    async with pool.acquire() as conn:
        row = await conn.fetchrow(sql, fsm_name, fsm_version, to_jsonb_param(state_json))
    return dict(row) if row else None


async def load_fsm_transition_from_json(
    pool: asyncpg.Pool, fsm_name: str, fsm_version: str, transition_json: Any
) -> Optional[dict]:
    """Mirrors loadFsmTransitionFromJsonV2() in fsm-helper.ts."""
    sql = f"SELECT * FROM {LOAD_FSM_TRANSITION_FN}($1::text, $2::text, $3::jsonb);"
    async with pool.acquire() as conn:
        row = await conn.fetchrow(sql, fsm_name, fsm_version, to_jsonb_param(transition_json))
    return dict(row) if row else None


async def load_fsm_from_json_v2(
    pool: asyncpg.Pool,
    fsm_name: str,
    fsm_version: str,
    fsm_json: Any,
    root_node_text: Optional[str] = None,
) -> Optional[dict]:
    """Calls fsm_core.load_fsm_from_json_v2 — orchestrates state + transition load with schema validation and caching."""
    sql = f"""
        SELECT {LOAD_FSM_FROM_JSON_FN}(
            $1::jsonb, $2::text, $3::text, $4::text
        ) AS result;
    """
    async with pool.acquire() as conn:
        row = await conn.fetchrow(
            sql, to_jsonb_param(fsm_json), root_node_text, fsm_name, fsm_version
        )
    if row and row["result"] is not None:
        result = row["result"]
        return dict(result) if hasattr(result, "keys") else result
    return None


async def resolve_state_value(
    pool: asyncpg.Pool, fsm_instance_id: str
) -> Optional[dict]:
    """Mirrors resolveStateValue() in fsm-helper.ts."""
    sql = f"SELECT {RESOLVE_STATE_VALUE_FN}($1::uuid) AS result;"
    async with pool.acquire() as conn:
        row = await conn.fetchrow(sql, fsm_instance_id)
    if row and row["result"] is not None:
        result = row["result"]
        return dict(result) if hasattr(result, "keys") else result
    return None

File: apps/fsm-core-db/src/fsm-helper.ts
import type { DBDeps } from "./custom-type.ts";
import { FSM_SCHEMA, FSM_SCHEMA_FN_VERSION } from "./const.ts";
import { toJsonbParam } from "./pg-utils.ts";


// TODO:  load_fsm_state_from_json_v2 fn

/**
 * Calls the load_fsm_state_from_json_v2 SQL function to load FSM state from JSON.
 * @param deps - DBDeps for database access
 * @param json_input - The FSM state JSONB input
 * @param root_node_text - The root node text (can be null)
 * @param fsm_name - The FSM name/identifier
 * @param fsm_version - The FSM version
 * @returns Promise<unknown> - The result from the SQL function
 */
export async function loadFsmStateFromJsonV2(
  deps: DBDeps,
  json_input: unknown,
  root_node_text: string | null,
  fsm_name: string,
  fsm_version: string,
): Promise<unknown> {
  try {
    const LOAD_FSM_STATE_FN = `${FSM_SCHEMA}.load_fsm_state_from_json_${FSM_SCHEMA_FN_VERSION}`;
    const text = `
      SELECT ${LOAD_FSM_STATE_FN}(
        $1::jsonb,
        $2::text,
        $3::text,
        $4::text
      ) AS result;
    `;
    const values = [
      toJsonbParam(json_input),
      root_node_text,
      fsm_name,
      fsm_version,
    ];
    const res = await deps.db.query<{ result: unknown }>(text, values);
    return res.rows?.[0]?.result ?? null;
  } catch (err) {
    console.error("Error in loadFsmStateFromJsonV2:", err);
    throw new Error("Failed to load FSM state from JSON", { cause: err });
  }
}

/**
 * Calls the load_fsm_transition_from_json_v2 SQL function to load FSM transitions from JSON.
 * @param deps - DBDeps for database access
 * @param json_input - The FSM transition JSONB input
 * @param root_node_text - The root node text (can be null)
 * @param fsm_name - The FSM name/identifier
 * @param fsm_version - The FSM version
 * @returns Promise<unknown> - The result from the SQL function
 */
export async function loadFsmTransitionFromJsonV2(
  deps: DBDeps,
  json_input: unknown,
  root_node_text: string | null,
  fsm_name: string,
  fsm_version: string,
): Promise<unknown> {
  try {
    const LOAD_FSM_TRANSITION_FN = `${FSM_SCHEMA}.load_fsm_transition_from_json_${FSM_SCHEMA_FN_VERSION}`;
    const text = `
      SELECT ${LOAD_FSM_TRANSITION_FN}(
        $1::jsonb,
        $2::text,
        $3::text,
        $4::text
      ) AS result;
    `;
    const values = [
      toJsonbParam(json_input),
      root_node_text,
      fsm_name,
      fsm_version,
    ];
    const res = await deps.db.query<{ result: unknown }>(text, values);
    return res.rows?.[0]?.result ?? null;
  } catch (err) {
    console.error("Error in loadFsmTransitionFromJsonV2:", err);
    throw new Error("Failed to load FSM transition from JSON", { cause: err });
  }
}

export async function loadFsmFromJsonV2(
  deps: DBDeps,
  json_input: unknown,
  root_node_text: string | null,
  fsm_name: string,
  fsm_version: string,
): Promise<unknown> {
  try {
    const LOAD_FSM_FROM_JSON_FN = `${FSM_SCHEMA}.load_fsm_from_json_${FSM_SCHEMA_FN_VERSION}`;
    const text = `
      SELECT ${LOAD_FSM_FROM_JSON_FN}(
        $1::jsonb,
        $2::text,
        $3::text,
        $4::text
      ) AS result;
    `;
    const values = [toJsonbParam(json_input), root_node_text, fsm_name, fsm_version];
    const res = await deps.db.query<{ result: unknown }>(text, values);
    return res.rows?.[0]?.result ?? null;
  } catch (err) {
    console.error("Error in loadFsmFromJsonV2:", err);
    throw new Error("Failed to load FSM from JSON", { cause: err });
  }
}

/**
 * Resolves state value for a given input JSON, FSM name, and FSM version.
 * Calls resolve_state_value_v2 SQL function and returns the result.
 * @param deps - DBDeps for database access
 * @param input_json - The state value JSONB to resolve
 * @param fsm_name - The FSM name/identifier
 * @param fsm_version - The FSM version
 * @returns Promise<{ json: unknown; all_nodes: string[] }> - The resolved state value and all nodes
 */
export async function resolveStateValue(
  deps: DBDeps,
  input_json: unknown,
  fsm_name: string,
  fsm_version: string,
): Promise<{ json: unknown; all_nodes: string[] } | null> {
  try {
    const RESOLVE_STATE_VALUE_FN = `${FSM_SCHEMA}.resolve_state_value_${FSM_SCHEMA_FN_VERSION}`;
    const text = `
      SELECT ${RESOLVE_STATE_VALUE_FN}(
        $1::jsonb,
        $2::text,
        $3::text
      ) AS result;
    `;
    const values = [
      toJsonbParam(input_json),
      fsm_name,
      fsm_version,
    ];
    const res = await deps.db.query<{
      result: { json: unknown; all_nodes: string[] };
    }>(text, values);
    if (!res.rows || res.rows.length === 0) return null;
    return res.rows[0]?.result ?? null;
  } catch (err) {
    console.error("Error in resolveStateValue:", err);
    throw new Error("Failed to resolve state value", { cause: err });
  }
}


export async function performMicrostep(
  deps: DBDeps,
  transition_record?: unknown,
  event_name: string,
  state_value_node_set: unknown,
  fsm_name: string,
  fsm_version: string,
): Promise<{
  updated_state_value_node_set: string[];
  updated_state_value: unknown;
  exit_actions: unknown;
  entry_actions: unknown;
  initial_actions: unknown;
  transition_actions: unknown;
}> {
  try {
    const transitionRecordJson =
      transition_record === undefined || transition_record === null
        ? null
        : JSON.stringify(transition_record);
    const stateValueNodeArray = Array.isArray(state_value_node_set)
      ? state_value_node_set
      : [];
    const MICROSTEP_FN = `${FSM_SCHEMA}.microstep_${FSM_SCHEMA_FN_VERSION}`;
    const text = `
      SELECT * FROM ${MICROSTEP_FN}(
        CASE
          WHEN $1::jsonb IS NULL THEN NULL::${FSM_SCHEMA}.fsm_transitions
          ELSE jsonb_populate_record(NULL::${FSM_SCHEMA}.fsm_transitions, $1::jsonb)
        END,
        $2::text,
        $3::text[],
        $4::text,
        $5::text
      ) AS result;
    `;
    const values = [
      transitionRecordJson,
      event_name,
      stateValueNodeArray,
      fsm_name,
      fsm_version,
    ];
    const res = await deps.db.query<{ result: {
      updated_state_value_node_set: string[];
      updated_state_value: unknown;
      exit_actions: unknown;
      entry_actions: unknown;
      initial_actions: unknown;
      transition_actions: unknown;
    } }>(text, values);
    return (
      res.rows?.[0]?.result ?? {
        updated_state_value_node_set: [],
        updated_state_value: null,
        exit_actions: null,
        entry_actions: null,
        initial_actions: null,
        transition_actions: null,
      }
    );
  } catch (err) {
    console.error("Error in performMicrostep:", err);
    throw new Error("Failed to perform microstep", { cause: err });
  }
}

// TODO:
// add fn with name: selectTransitions that will call select_all_transitions
/**
 * Fetches all transitions for a given event, FSM name, and version by calling the SQL function select_all_transitions.
 * @param deps - DBDeps for database access
 * @param event_name - The event name to filter transitions
 * @param source_state_value_set - The source state value set to filter transitions
 * @param fsm_name - The FSM name/identifier
 * @param fsm_version - The FSM version
 * @returns Promise<Array<any>> - Array of transition records
 */
export async function selectTransitions(
  deps: DBDeps,
  event_name: string,
  source_state_value_set: unknown,
  fsm_name: string,
  fsm_version: string,
): Promise<any[]> {
  try {
    const SELECT_TRANSITIONS_FN = `${FSM_SCHEMA}.select_all_transitions_${FSM_SCHEMA_FN_VERSION}`;
    const text = `
      SELECT * FROM ${SELECT_TRANSITIONS_FN}(
        $1::text,
        $2::jsonb,
        $3::text,
        $4::text
      );
    `;
    const values = [
      event_name,
      toJsonbParam(source_state_value_set),
      fsm_name,
      fsm_version,
    ];
    const res = await deps.db.query(text, values);
    return res.rows ?? [];
  } catch (err) {
    console.error("Error in selectTransitions:", err);
    throw new Error("Failed to select transitions", { cause: err });
  }
}
